File: src/apps/daemon/internal/networking/proxyconfig/subscription_crawler_pipeline.go
package proxyconfig

import (
	"encoding/base64"
	"sort"
	"strings"
)

type CrawlSource struct {
	URL            string `json:"url"`
	IntervalSecs   int64  `json:"interval_secs"`
	LastCrawl      int64  `json:"last_crawl"`
	TotalHarvested int    `json:"total_harvested"`
	Enabled        bool   `json:"enabled"`
}

type SubscriptionCrawlerPipeline struct {
	sources   map[string]*CrawlSource
	harvested map[string]struct{}
}

func NewSubscriptionCrawlerPipeline() *SubscriptionCrawlerPipeline {
	return &SubscriptionCrawlerPipeline{
		sources:   make(map[string]*CrawlSource),
		harvested: make(map[string]struct{}),
	}
}

func (s *SubscriptionCrawlerPipeline) AddSource(url string, intervalSecs int64) {
	if intervalSecs < 60 {
		intervalSecs = 60
	}
	s.sources[url] = &CrawlSource{
		URL:          url,
		IntervalSecs: intervalSecs,
		LastCrawl:    0,
		Enabled:      true,
	}
}

func (s *SubscriptionCrawlerPipeline) DispatchPendingSources(currentTime int64) []string {
	var toCrawl []string
	for _, src := range s.sources {
		if src.Enabled && (src.LastCrawl == 0 || currentTime >= src.LastCrawl+src.IntervalSecs) {
			src.LastCrawl = currentTime
			toCrawl = append(toCrawl, src.URL)
		}
	}
	return toCrawl
}

func (s *SubscriptionCrawlerPipeline) IngestCrawlContent(sourceURL, rawContent string) int {
	trimmed := strings.TrimSpace(rawContent)
	decodedBytes, err := base64.StdEncoding.DecodeString(trimmed)
	decodedText := trimmed
	if err == nil {
		decodedText = string(decodedBytes)
	}

	count := 0
	lines := strings.Split(decodedText, "\n")
	for _, line := range lines {
		l := strings.TrimSpace(line)
		if strings.HasPrefix(l, "ss://") ||
			strings.HasPrefix(l, "vmess://") ||
			strings.HasPrefix(l, "vless://") ||
			strings.HasPrefix(l, "trojan://") ||
			strings.HasPrefix(l, "hysteria2://") ||
			strings.HasPrefix(l, "tuic://") {
			if _, exists := s.harvested[l]; !exists {
				s.harvested[l] = struct{}{}
				count++
			}
		}
	}

	if src, ok := s.sources[sourceURL]; ok {
		src.TotalHarvested += count
	}

	return count
}

func (s *SubscriptionCrawlerPipeline) GetHarvestedProxies() []string {
	var list []string
	for p := range s.harvested {
		list = append(list, p)
	}
	sort.Strings(list)
	return list
}

func (s *SubscriptionCrawlerPipeline) TotalHarvestedCount() int {
	return len(s.harvested)
}

File: src/packages/lumicore/src/proxy/subscription_crawler_pipeline.rs
//! # Subscription Crawler Pipeline
//!
//! Multi-source subscription crawling pipeline, scheduled fetch dispatcher,
//! base64/plain payload extractor, and proxy candidate deduplication filter.

use base64::{engine::general_purpose::STANDARD as BASE64_STANDARD, Engine as _};
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CrawlSource {
    pub url: String,
    pub interval_secs: u64,
    pub last_crawl: u64,
    pub total_harvested: usize,
    pub enabled: bool,
}

pub struct SubscriptionCrawlerPipeline {
    sources: HashMap<String, CrawlSource>,
    harvested_proxies: HashSet<String>,
}

impl SubscriptionCrawlerPipeline {
    pub fn new() -> Self {
        Self {
            sources: HashMap::new(),
            harvested_proxies: HashSet::new(),
        }
    }

    pub fn add_source(&mut self, url: &str, interval_secs: u64) {
        self.sources.insert(
            url.to_string(),
            CrawlSource {
                url: url.to_string(),
                interval_secs: interval_secs.max(60),
                last_crawl: 0,
                total_harvested: 0,
                enabled: true,
            },
        );
    }

    pub fn dispatch_pending_sources(&mut self, current_time: u64) -> Vec<String> {
        let mut to_crawl = Vec::new();
        for source in self.sources.values_mut() {
            if source.enabled && (source.last_crawl == 0 || current_time >= source.last_crawl + source.interval_secs) {
                source.last_crawl = current_time;
                to_crawl.push(source.url.clone());
            }
        }
        to_crawl
    }

    pub fn ingest_crawl_content(&mut self, source_url: &str, raw_content: &str) -> usize {
        let mut count = 0;
        let content_trimmed = raw_content.trim();

        // Check if entire payload is base64 encoded
        let decoded_text = if let Ok(decoded_bytes) = BASE64_STANDARD.decode(content_trimmed) {
            String::from_utf8(decoded_bytes).unwrap_or_else(|_| content_trimmed.to_string())
        } else {
            content_trimmed.to_string()
        };

        for line in decoded_text.lines() {
            let line_trimmed = line.trim();
            if line_trimmed.starts_with("ss://")
                || line_trimmed.starts_with("vmess://")
                || line_trimmed.starts_with("vless://")
                || line_trimmed.starts_with("trojan://")
                || line_trimmed.starts_with("hysteria2://")
                || line_trimmed.starts_with("tuic://")
            {
                if self.harvested_proxies.insert(line_trimmed.to_string()) {
                    count += 1;
                }
            }
        }

        if let Some(source) = self.sources.get_mut(source_url) {
            source.total_harvested += count;
        }

        count
    }

    pub fn get_harvested_proxies(&self) -> Vec<String> {
        let mut list: Vec<_> = self.harvested_proxies.iter().cloned().collect();
        list.sort();
        list
    }

    pub fn total_harvested_count(&self) -> usize {
        self.harvested_proxies.len()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_crawler_pipeline() {
        let mut pipeline = SubscriptionCrawlerPipeline::new();
        pipeline.add_source("https://example.com/subs.txt", 3600);

        let pending = pipeline.dispatch_pending_sources(1000);
        assert_eq!(pending.len(), 1);
        assert_eq!(pending[0], "https://example.com/subs.txt");

        // No new crawls within interval
        let pending_too_soon = pipeline.dispatch_pending_sources(1500);
        assert!(pending_too_soon.is_empty());

        let raw = "vmess://eyJhZGRyIjoiMS4xLjEuMSJ9\ntrojan://pass@2.2.2.2:443\ninvalid_line";
        let ingested = pipeline.ingest_crawl_content("https://example.com/subs.txt", raw);
        assert_eq!(ingested, 2);
        assert_eq!(pipeline.total_harvested_count(), 2);
    }

    #[test]
    fn test_base64_encoded_crawl_content() {
        let mut pipeline = SubscriptionCrawlerPipeline::new();
        pipeline.add_source("https://example.com/b64", 3600);

        let lines = "ss://user:pass@1.2.3.4:8388\nvless://uuid@5.6.7.8:443";
        let encoded = BASE64_STANDARD.encode(lines);

        let ingested = pipeline.ingest_crawl_content("https://example.com/b64", &encoded);
        assert_eq!(ingested, 2);
        assert_eq!(pipeline.total_harvested_count(), 2);
    }
}
